src/Proto.Actor/Futures.cs

```diff
@@ -1,33 +1,39 @@
-﻿// -----------------------------------------------------------------------
-//  <copyright file="Futures.cs" company="Asynkron HB">
-//      Copyright (C) 2015-2017 Asynkron HB All rights reserved
-//  </copyright>
-// -----------------------------------------------------------------------
-
-using System.Threading.Tasks;
-
-namespace Proto
-{
-    public class FutureProcess<T> : Process
-    {
-        private readonly TaskCompletionSource<T> _tcs;
-
-        public FutureProcess(TaskCompletionSource<T> tcs)
-        {
-            _tcs = tcs;
-        }
-
-        public override void SendUserMessage(PID pid, object message, PID sender)
-        {
-            if (message is T)
-            {
-                _tcs.TrySetResult((T) message);
-                pid.Stop();
-            }
-        }
-
-        public override void SendSystemMessage(PID pid, object message)
-        {
-        }
-    }
+﻿// -----------------------------------------------------------------------
+//  <copyright file="Futures.cs" company="Asynkron HB">
+//      Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Proto
+{
+    public class FutureProcess<T> : Process
+    {
+        private readonly TaskCompletionSource<T> _tcs;
+        private readonly CancellationToken _cancellationToken;
+
+        public FutureProcess(TaskCompletionSource<T> tcs) : this(tcs, CancellationToken.None) { }
+        public FutureProcess(TaskCompletionSource<T> tcs, CancellationToken cancellationToken)
+        {
+            _tcs = tcs;
+            _cancellationToken = cancellationToken;
+        }
+
+        public override void SendUserMessage(PID pid, object message, PID sender)
+        {
+            if (message is T)
+            {
+                if (_cancellationToken.IsCancellationRequested) return;
```

Comment:
How about if we move the CancellationTokenSource and the logic related to it into the FutureProcess. Then we don't clutter the PID with all of the timeout/cancellation logic. The only difference in PID would be that we optionally pass a CancellationTokenSource to the FutureProcess, and we would return FutureProcess.Task instead of TaskCompletionSource.Task. Does that make sense? :)